caikit/runtime/utils/servicer_util.py

```diff
@@ -260,6 +269,140 @@ def validate_data_model(
+            self._stream.__class__.ELEMENT_TYPE,
+            self._stream.__getstate__(),
+        )
+
+    def __setstate__(self, pickled: Tuple[str, str, type, bytes]):
+        """Unpickle the values from pickling and fall back to dynamic service
+        generation if needed.
+
+        Args:
+            pickled (Tuple[str, str, type, bytes]): The four elements created
+                when pickling this instance
+        """
+        stream_type_module, stream_type_name, element_type_class, stream_bytes = pickled
+        try:
+            mod = sys.modules[stream_type_module]
+            stream_type = getattr(mod, stream_type_name)
+        except (KeyError, AttributeError):
+            # Local
+            from caikit.runtime.service_factory import (  # pylint: disable=import-outside-toplevel
+                ServicePackageFactory,
+            )
+
+            ServicePackageFactory.get_service_package(
+                ServicePackageFactory.ServiceType.INFERENCE
+            )
+            ServicePackageFactory.get_service_package(
+                ServicePackageFactory.ServiceType.TRAINING
+            )
+            stream_type = make_data_stream_source(element_type_class)
+        self._stream = stream_type.__new__(stream_type)
+        self._stream.__setstate__(stream_bytes)
+
+    def __getattr__(self, name: str) -> Any:
+        """Forward all getattr requests to the underlying stream"""
+        return getattr(self._stream, name)
+
+
+class SpawnProcessModelWrapper(ModuleBase):
+    """This class wraps up a model to make it safe to pass to a spawned
+    subprocess. It will not be efficient, but it will be safe!
+    """
+
+    def __init__(self, model_id: str, model: ModuleBase):
+        super().__init__()
+        self._model_id = model_id
+        self._model = model
+
+    def __getattr__(self, name):
+        """Forward attributes that are not found on the base class to the model
+
+        NOTE: This does _not_ forward base class attributes since those are
+            resolved before __getattr__ is called.
+        """
+        return getattr(self._model, name)
+
+    def save(self, *args, **kwargs):
+        """Directly forward save to the model so that it is not called by the
+        base class
+        """
+        return self._model.save(*args, **kwargs)
+
+    def run(self, *args, **kwargs):
+        """Directly forward run to the model so that it is not called by the
+        base class
+        """
+        return self._model.run(*args, **kwargs)
+
+    def __getstate__(self) -> Tuple[str, Optional[bytes]]:
+        """When pickling, only send the serialized model body for non-fork. This
+        is not a general-purpose pickle solution for models, but makes them safe
+        for training jobs that need to move models between processes.
+        """
+        start_method = multiprocessing.get_start_method()
+        body = None
+        if start_method != "fork":
+            body = self._model.as_bytes()
```

Comment:
nice!

Though this also requires that models implement `.save()` which they might not